src/lcos/detail/future_data.cpp

```diff
@@ -237,31 +239,48 @@ namespace hpx { namespace lcos { namespace detail
     {
         if (!data_sink) return;
 
-        std::unique_lock<mutex_type> l(mtx_);
-
-        if (is_ready_locked(l)) {
-
-            HPX_ASSERT(on_completed_.empty());
-
+        if (is_ready())
+        {
             // invoke the callback (continuation) function right away
-            l.unlock();
-
             handle_on_completed(std::move(data_sink));
         }
-        else {
-            on_completed_.push_back(std::move(data_sink));
+        else
+        {
+            std::unique_lock<mutex_type> l(mtx_);
+            if (is_ready())
+            {
+#if !(defined(HPX_DISABLE_ASSERTS) || defined(BOOST_DISABLE_ASSERTS) ||        \
+    defined(NDEBUG))
+                // HPX_ASSERT might suspend
+                auto t = std::move(on_completed_);
+                l.unlock();
+                HPX_ASSERT(t.empty());
```

Comment:
for this condition to hold, setting the state to ready needs to be done under the lock.

Reply:
Good catch... I think we don't need to verify this condition in the first place.